sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala

```diff
@@ -26,7 +26,7 @@ import org.apache.spark.sql.types.Decimal.DecimalIsConflicted
 private[sql] object ByteExactNumeric extends ByteIsIntegral with Ordering.ByteOrdering {
   private def checkOverflow(res: Int, x: Byte, y: Byte, op: String): Unit = {
     if (res > Byte.MaxValue || res < Byte.MinValue) {
-      throw new ArithmeticException(s"$x $op $y caused overflow.")
```

Comment:
As an end user, I would prefer to see which particular values caused the issue. Unification is nice but it shouldn't make user experience worse, IMHO.

Reply:
Yes, I thought about this. 
On the other way, we can align the error message of int/log to the error message of byte/short, which is more user-friendly. Then basically we have to re-implement the "exact" methods in Spark.
cc @cloud-fan

Reply:
It's a bit risky to implement the "exact" methods ourselves, as JDK may update them in future versions. I'd rather add a try-catch to change the error message.

Reply:
> It's a bit risky to implement the "exact" methods ourselves, as JDK may update them in future versions. I'd rather add a try-catch to change the error message.

So which one do you prefer? 
1. Changing the error message of byte/short overflow as simply: "tinyint/smallint overflow" (I check PostgreSQL and it  also simply show error messages like "ERROR: integer out of range")
2.Add try/catch in int/long arithmetic operations and throw a new exception with details.

cc @MaxGekk @cloud-fan @maropu

Reply:
Actually, we have > 100 places where we use `*Exact` methods from the standard library. Such methods just throw the exception w/o any clues about inputs:
```java
    public static int multiplyExact(int x, int y) {
        long r = (long)x * (long)y;
        if ((int)r != r) {
            throw new ArithmeticException("integer overflow");
        }
        return (int)r;
    }
```
As the first step, we could align our `ArithmeticException` to the standard library (just a few places). After that we can think of how we can improve such kind of exceptions in all possible places in Spark, and give users more context.

Reply:
> For Int/Long, the message is "int/long overflow" since Spark is calling the "*Exact"(e.g. addExact, negateExact) methods from java.lang.Math.

The unification itself looks nice and I think we should add try-catch in the int/log cases first.